hugegraph-core/src/main/java/com/baidu/hugegraph/cluster/ServerInfoManager.java

```diff
@@ -0,0 +1,247 @@
+
+import static com.baidu.hugegraph.backend.page.PageInfo.PAGE_NONE;
+import static com.baidu.hugegraph.backend.query.Query.NO_LIMIT;
+
+public class ServerInfoManager {
+
+    private static final Logger LOG = Log.logger(ServerInfoManager.class);
+
+    private final HugeGraphParams graph;
+    private final EventListener eventListener;
+    private Id serverId;
+    private NodeRole serverRole;
+
+    public ServerInfoManager(HugeGraphParams graph) {
+        E.checkNotNull(graph, "graph");
+        this.graph = graph;
+        this.eventListener = this.listenChanges();
+    }
+
+    private EventListener listenChanges() {
+        // Listen store event: "store.inited"
+        Set<String> storeEvents = ImmutableSet.of(Events.STORE_INITED);
+        EventListener eventListener = event -> {
+            // Ensure user schema create after system info initialized
+            if (storeEvents.contains(event.name())) {
+                try {
+                    this.initSchemaIfNeeded();
+                } finally {
+                    this.graph.closeTx();
+                }
+                return true;
+            }
+            return false;
+        };
+        this.graph.loadSystemStore().provider().listen(eventListener);
+        return eventListener;
+    }
+
+    private void unlistenChanges() {
+        this.graph.loadSystemStore().provider().unlisten(this.eventListener);
+    }
+
+    public boolean close() {
+        this.unlistenChanges();
+        return true;
+    }
+
+    public void initServerInfo(String server, String role) {
+        E.checkArgument(server != null && !server.isEmpty(),
+                        "The server name can't be null or empty");
+        HugeServerInfo existed = this.serverInfo(server);
+        E.checkArgument(existed == null || !existed.alive(),
+                        "The server with name '%s' already in cluster",
+                        server);
+        E.checkArgument(role != null && !role.isEmpty(),
+                        "The server role can't be null or empty");
+        NodeRole nodeRole = NodeRole.valueOf(role.toUpperCase());
+        if (nodeRole.master()) {
+            String page = PAGE_NONE;
+            do {
+                Iterator<HugeServerInfo> servers = this.serverInfos(10L, page);
+                while (servers.hasNext()) {
+                    existed = servers.next();
+                    E.checkArgument(existed.role().worker() ||
+                                    !existed.alive(),
+                                    "Already existed master '%s' in current " +
+                                    "cluster", existed.id());
+                }
+                page = PageInfo.pageInfo(servers);
+            } while (page != null);
+        }
+
+        HugeServerInfo serverInfo = new HugeServerInfo(server, nodeRole);
+        serverInfo.maxLoad(this.calcMaxLoad());
+        this.serverId = serverInfo.id();
+        this.serverRole = nodeRole;
```

Comment:
move serverId and serverRole assignment to the begin of this method after refactor initServerInfo params